### PR TITLE
Request object features per plugin in parallel

### LINKED ISSUE #2744: Rework ObjectFeaturesPlugin::compute_local
To support easy writing of stable and fast plugins for object feature computation, the bones exist in the ObjectFeaturesPlugin::compute_local method, but writing is hampered by some implementation details.

- By default this method gives the neighborhood along with the object, this should be another method or settable behaviour
- The compute_local loop is unparallelized and not integrated into the request architecture of ilastik multithreading
- compute_local does not get basic object information such as object_id and/or location, which make debugging difficult
- good example features that are not connected to vigra (which adds historical complexity) are not in the codebase

If these would be fixed, it opens up a very easy route to make any object quantifier directly attached to ilastik as a new one would generally only require duplicating the example and changing the name of the feature and the compute_local function, meaning that ilastik would be a viable place for any quantitative object extraction that you may want to pipe into a classifier.

### PR DESCRIPTION
**Pull request by k-dominik**
Use request system to allow for parallel computation of local object features.

Will speed up computation if gil is released in features `compute_local` somewhere.

fixes part of #2744 

Todos:

- [x] cleanup
- [x] benchmark our features (also with time-series)
   for our [drosophila example](https://data.ilastik.org/drosophila.zip) I found a 2-3 times speedup in feature computation (all excl location) on my M1 without any numbers changing.

### CODE REVIEW

File: ilastik/applets/objectExtraction/opObjectExtraction.py [hunk 1]
```diff
@@ -708,19 +715,37 @@ def dictextend(a, b):
...
+                        bboxes[i] = (raw_bbox, binary_bbox)
+
+                    raw_bbox, binary_bbox = bboxes[i]
+                    pool.add(Request(partial(_calc_single, i, raw_bbox, binary_bbox)))
+
+                pool.wait()
+                pool.clean()
+                # merge the results
+                for i in range(0, nobj):
+                    local_features[plugin_name] = dictextend(local_features[plugin_name], tmp_dicts[i])
```
Comment (emilmelnikov) [inline ilastik/applets/objectExtraction/opObjectExtraction.py:748]:
```suggestion
                for feature_dict in tmp_dicts:
                    for feature_name, features in feature_dict.items():
                        local_features[plugin_name][feature_name].append(features)
```

...and there is no need for `dictextend`.

**Review by emilmelnikov — COMMENTED** [review]
It's out of the scope of this PR, but this method could use some refactoring.

File: ilastik/applets/objectExtraction/opObjectExtraction.py (hunk 1 above)
Comment (k-dominik) [reply ilastik/applets/objectExtraction/opObjectExtraction.py:748]: Hm, yeah, you could inline it. I'm not sure it's better for readability here. Do you think so?

Comment (emilmelnikov) [reply ilastik/applets/objectExtraction/opObjectExtraction.py:748]: It is used only in one place, the name doesn't explain what it actually does (at least to me), and the double loop has `dict1[key1][key2].append(value)` statement that shows how the main data structure is updated.

File: ilastik/applets/objectExtraction/opObjectExtraction.py [hunk 2]
```diff
@@ -708,19 +715,37 @@ def dictextend(a, b):
...
+                        extent = self.compute_extent(i, image, mincoords, maxcoords, axes, margin)
+                        raw_bbox = self.compute_rawbbox(image, extent, axes)
+                        # it's i+1 here, because the background has label 0
+                        binary_bbox = numpy.where(labels[tuple(extent)] == i + 1, 1, 0).astype(bool)
+                        bboxes[i] = (raw_bbox, binary_bbox)
+
+                    raw_bbox, binary_bbox = bboxes[i]
+                    pool.add(Request(partial(_calc_single, i, raw_bbox, binary_bbox)))
+
+                pool.wait()
```
[the remark this answers is not in the document]
Comment (emilmelnikov) [reply ilastik/applets/objectExtraction/opObjectExtraction.py:744]: The following approach could also work if there is no need to cancel (the name could be better though):

```python
def run_in_pool(funcs: Sequence[Callable]) -> None:
    pool = RequestPool()
    for func in funcs:
        pool.add(Request(func))
    pool.wait()


funcs = []
for something in sequence:
    ...
    funcs.append(partial(func, *args))
    ...

run_in_pool(funcs)
```

[the remark this answers is not in the document]
Comment (k-dominik) [reply ilastik/applets/objectExtraction/opObjectExtraction.py]: also a nice pattern, but I kinda like using a contextmanager. It is a bit closer to concurrent.futures...

**Comment by k-dominik** [comment]
merging this, there were no major changes after the review, mostly CI gymnastics...

**Pull request merged**